sway-lsp/src/global_state.rs

```diff
@@ -0,0 +1,176 @@
+            let mut diagnostics_to_publish = vec![];
+            let config = &self.config.read();
+            let engines = session.engines.read();
+            let tokens = session.token_map().tokens_for_file(engines.se(), uri);
+            match config.debug.show_collected_tokens_as_warnings {
+                // If collected_tokens_as_warnings is Parsed or Typed,
+                // take over the normal error and warning display behavior
+                // and instead show the either the parsed or typed tokens as warnings.
+                // This is useful for debugging the lsp parser.
+                Warnings::Parsed => diagnostics_to_publish
+                    .extend(debug::generate_warnings_for_parsed_tokens(tokens)),
+                Warnings::Typed => {
+                    diagnostics_to_publish.extend(debug::generate_warnings_for_typed_tokens(tokens))
+                }
+                Warnings::Default => {}
+            }
+            let diagnostics = session.wait_for_parsing();
+            if config.diagnostic.show_warnings {
+                diagnostics_to_publish.extend(diagnostics.warnings);
+            }
+            if config.diagnostic.show_errors {
+                diagnostics_to_publish.extend(diagnostics.errors);
+            }
+            diagnostics_to_publish
+        };
+
+        // Note: Even if the computed diagnostics vec is empty, we still have to push the empty Vec
+        // in order to clear former diagnostics. Newly pushed diagnostics always replace previously pushed diagnostics.
+        self.client
+            .publish_diagnostics(workspace_uri.clone(), diagnostics_res, None)
+            .await;
+    }
+
+    pub(crate) async fn parse_project(&self, uri: Url, workspace_uri: Url, session: Arc<Session>) {
+        let should_publish = run_blocking_parse_project(uri.clone(), session.clone()).await;
+        if should_publish {
+            self.publish_diagnostics(&uri, &workspace_uri, session)
+                .await;
+        }
+    }
+}
+
+/// Runs parse_project in a blocking thread, because parsing is not async.
+async fn run_blocking_parse_project(uri: Url, session: Arc<Session>) -> bool {
+    task::spawn_blocking(move || match session.parse_project(&uri) {
+        Ok(should_publish) => should_publish,
+        Err(err) => {
+            tracing::error!("{}", err);
+            matches!(err, LanguageServerError::FailedToParse)
+        }
+    })
+    .await
+    .unwrap_or_default()
+}
+
+/// `Sessions` is a collection of `Session`s, each of which represents a project
+/// that has been opened in the users workspace.
+pub(crate) struct Sessions(DashMap<PathBuf, Arc<Session>>);
+
+impl Sessions {
+    fn init(&self, uri: &Url) -> Result<(), LanguageServerError> {
+        let session = Arc::new(Session::new());
+        let project_name = session.init(uri)?;
+        self.insert(project_name, session);
+        Ok(())
+    }
+
+    pub(crate) fn get_uri_and_session(
```

Comment:
nit: a better name for this function could be something like `from_workspace_uri` - something to indicate it's pulling everything you need from Sessions given a workspace URI.

Reply:
Nice, i'm happy to make this change. Feels more idiomatic.